File: sdk/src/queries/financials.ts
import type { DataPlatClient } from '../client.js';
import type {
  Financial,
  FinancialsParams,
  MetricParams,
  MetricDataPoint,
} from '../types.js';

function escTicker(t: string): string {
  return t.replace(/'/g, "\\'");
}

function tickerList(tickers: string[]): string {
  return tickers.map((t) => `'${escTicker(t)}'`).join(', ');
}

/**
 * Map the timeframe param ('annual'|'quarterly') to a SQL filter
 * using the new schema's fiscal_period column.
 *   annual     → fiscal_period = 'FY'
 *   quarterly  → fiscal_period IN ('Q1','Q2','Q3','Q4')
 */
function timeframeFilter(tf?: 'annual' | 'quarterly'): string {
  if (tf === 'annual') return "fiscal_period = 'FY'";
  if (tf === 'quarterly') return "fiscal_period IN ('Q1','Q2','Q3','Q4')";
  return '';
}

/** Derive a virtual timeframe value from fiscal_period for backward compat. */
function deriveTimeframe(fiscalPeriod: string): 'annual' | 'quarterly' {
  return fiscalPeriod === 'FY' ? 'annual' : 'quarterly';
}

interface RawFinancialRow {
  ticker: string;
  cik: string | null;
  period_start: string | null;
  period_end: string;
  fiscal_year: string;
  fiscal_period: string;
  form_type: string;
  filed_date: string | null;
  accession_number: string | null;

  // Income statement
  revenue: number | null;
  cost_of_revenue: number | null;
  gross_profit: number | null;
  operating_expenses: number | null;
  operating_income: number | null;
  net_income: number | null;
  basic_eps: number | null;
  diluted_eps: number | null;
  research_and_dev: number | null;
  sga_expenses: number | null;
  income_tax: number | null;
  interest_expense: number | null;
  ebitda: number | null;

  // Balance sheet
  total_assets: number | null;
  current_assets: number | null;
  noncurrent_assets: number | null;
  total_liabilities: number | null;
  current_liabilities: number | null;
  noncurrent_liabilities: number | null;
  total_equity: number | null;
  retained_earnings: number | null;
  long_term_debt: number | null;
  short_term_debt: number | null;
  cash_and_equivalents: number | null;
  inventory: number | null;
  accounts_receivable: number | null;
  accounts_payable: number | null;
  goodwill: number | null;

  // Cash flow
  operating_cash_flow: number | null;
  investing_cash_flow: number | null;
  financing_cash_flow: number | null;
  capex: number | null;
  dividends_paid: number | null;
  depreciation_amortization: number | null;

  // Dilution & shares
  shares_outstanding: number | null;
  weighted_avg_shares_basic: number | null;
  weighted_avg_shares_diluted: number | null;
  stock_based_compensation: number | null;
  buyback_value: number | null;
  dividends_per_share: number | null;
}

function mapRow(r: RawFinancialRow): Financial {
  return {
    ticker: r.ticker,
    cik: r.cik,
    periodStart: r.period_start,
    periodEnd: r.period_end,
    fiscalYear: r.fiscal_year,
    fiscalPeriod: r.fiscal_period,
    formType: r.form_type,
    filedDate: r.filed_date,
    accessionNumber: r.accession_number,

    revenue: r.revenue,
    costOfRevenue: r.cost_of_revenue,
    grossProfit: r.gross_profit,
    operatingExpenses: r.operating_expenses,
    operatingIncome: r.operating_income,
    netIncome: r.net_income,
    basicEps: r.basic_eps,
    dilutedEps: r.diluted_eps,
    researchAndDev: r.research_and_dev,
    sgaExpenses: r.sga_expenses,
    incomeTax: r.income_tax,
    interestExpense: r.interest_expense,
    ebitda: r.ebitda,

    totalAssets: r.total_assets,
    currentAssets: r.current_assets,
    noncurrentAssets: r.noncurrent_assets,
    totalLiabilities: r.total_liabilities,
    currentLiabilities: r.current_liabilities,
    noncurrentLiabilities: r.noncurrent_liabilities,
    totalEquity: r.total_equity,
    retainedEarnings: r.retained_earnings,
    longTermDebt: r.long_term_debt,
    shortTermDebt: r.short_term_debt,
    cashAndEquivalents: r.cash_and_equivalents,
    inventory: r.inventory,
    accountsReceivable: r.accounts_receivable,
    accountsPayable: r.accounts_payable,
    goodwill: r.goodwill,

    operatingCashFlow: r.operating_cash_flow,
    investingCashFlow: r.investing_cash_flow,
    financingCashFlow: r.financing_cash_flow,
    capex: r.capex,
    dividendsPaid: r.dividends_paid,
    depreciationAmortization: r.depreciation_amortization,

    sharesOutstanding: r.shares_outstanding,
    weightedAvgSharesBasic: r.weighted_avg_shares_basic,
    weightedAvgSharesDiluted: r.weighted_avg_shares_diluted,
    stockBasedCompensation: r.stock_based_compensation,
    buybackValue: r.buyback_value,
    dividendsPerShare: r.dividends_per_share,

    // Derived for backward compat
    timeframe: deriveTimeframe(r.fiscal_period),
  };
}

export async function getFinancials(
  client: DataPlatClient,
  params: FinancialsParams,
): Promise<Financial[]> {
  const filters: string[] = [`ticker = '${escTicker(params.ticker)}'`];

  const tfFilter = timeframeFilter(params.timeframe);
  if (tfFilter) filters.push(tfFilter);

  if (params.fiscalYear) {
    filters.push(`fiscal_year = '${params.fiscalYear}'`);
  }
  if (params.fiscalPeriod) {
    filters.push(`fiscal_period = '${params.fiscalPeriod}'`);
  }

  const limit = params.limit ?? 40;

  const sql = `
    SELECT *
    FROM financials
    WHERE ${filters.join(' AND ')}
    ORDER BY period_end DESC
    LIMIT ${limit}
  `;

  const result = await client.query<RawFinancialRow>(sql);
  return result.rows.map(mapRow);
}

export async function getIncomeStatement(
  client: DataPlatClient,
  ticker: string,
  timeframe: 'annual' | 'quarterly' = 'quarterly',
  limit = 12,
): Promise<Financial[]> {
  return getFinancials(client, { ticker, timeframe, limit });
}

export async function getBalanceSheet(
  client: DataPlatClient,
  ticker: string,
  timeframe: 'annual' | 'quarterly' = 'quarterly',
  limit = 12,
): Promise<Financial[]> {
  return getFinancials(client, { ticker, timeframe, limit });
}

export async function getCashFlow(
  client: DataPlatClient,
  ticker: string,
  timeframe: 'annual' | 'quarterly' = 'quarterly',
  limit = 12,
): Promise<Financial[]> {
  return getFinancials(client, { ticker, timeframe, limit });
}

/** Get a single metric across multiple tickers for comparison. */
export async function getMetric(
  client: DataPlatClient,
  params: MetricParams,
): Promise<MetricDataPoint[]> {
  if (params.tickers.length === 0) return [];

  // Map camelCase metric name to snake_case column
  const colMap: Record<string, string> = {
    revenue: 'revenue',
    costOfRevenue: 'cost_of_revenue',
    grossProfit: 'gross_profit',
    operatingExpenses: 'operating_expenses',
    operatingIncome: 'operating_income',
    netIncome: 'net_income',
    basicEps: 'basic_eps',
    dilutedEps: 'diluted_eps',
    totalAssets: 'total_assets',
    totalLiabilities: 'total_liabilities',
    totalEquity: 'total_equity',
    longTermDebt: 'long_term_debt',
    operatingCashFlow: 'operating_cash_flow',
    investingCashFlow: 'investing_cash_flow',
    financingCashFlow: 'financing_cash_flow',
    researchAndDev: 'research_and_dev',
    sgaExpenses: 'sga_expenses',
    ebitda: 'ebitda',
    interestExpense: 'interest_expense',
    capex: 'capex',
    sharesOutstanding: 'shares_outstanding',
    weightedAvgSharesBasic: 'weighted_avg_shares_basic',
    weightedAvgSharesDiluted: 'weighted_avg_shares_diluted',
    stockBasedCompensation: 'stock_based_compensation',
    cashAndEquivalents: 'cash_and_equivalents',
    retainedEarnings: 'retained_earnings',
    shortTermDebt: 'short_term_debt',
  };

  const col = colMap[params.metric as string] ?? (params.metric as string);
  const limit = params.limit ?? 20;

  const tfFilter = timeframeFilter(params.timeframe);
  const tfClause = tfFilter ? `AND ${tfFilter}` : '';

  const sql = `
    SELECT
      ticker,
      toString(period_end) AS periodEnd,
      fiscal_year AS fiscalYear,
      fiscal_period AS fiscalPeriod,
      ${col} AS value
    FROM financials
    WHERE ticker IN (${tickerList(params.tickers)})
      ${tfClause}
    ORDER BY ticker, period_end DESC
    LIMIT ${limit * params.tickers.length}
  `;

  const result = await client.query<MetricDataPoint>(sql);
  return result.rows;
}

File: src/dataplat/ingestion/edgar/financials.py
"""SEC EDGAR financials pipeline — companyfacts → financials table."""

from __future__ import annotations

import logging
import time

import httpx
import polars as pl

from dataplat.db.client import get_client
from dataplat.db.migrate import ensure_schema
from dataplat.ingestion.edgar.cik_map import CIKMap
from dataplat.ingestion.edgar.client import build_filing_url, get_companyfacts, make_client
from dataplat.ingestion.edgar.concepts import extract_financials

logger = logging.getLogger(__name__)

# All columns in the financials table (must match migration 019)
FINANCIALS_COLUMNS = [
    "ticker", "cik", "period_start", "period_end", "fiscal_year", "fiscal_period",
    "form_type", "filed_date", "accession_number",
    # Income
    "revenue", "cost_of_revenue", "gross_profit", "operating_expenses",
    "operating_income", "net_income", "basic_eps", "diluted_eps",
    "research_and_dev", "sga_expenses", "income_tax", "interest_expense", "ebitda",
    # Balance
    "total_assets", "current_assets", "noncurrent_assets",
    "total_liabilities", "current_liabilities", "noncurrent_liabilities",
    "total_equity", "retained_earnings", "long_term_debt", "short_term_debt",
    "cash_and_equivalents", "inventory", "accounts_receivable", "accounts_payable", "goodwill",
    # Cash Flow
    "operating_cash_flow", "investing_cash_flow", "financing_cash_flow",
    "capex", "dividends_paid", "depreciation_amortization",
    # Dilution
    "shares_outstanding", "shares_issued",
    "weighted_avg_shares_basic", "weighted_avg_shares_diluted",
    "stock_based_compensation", "buyback_shares", "buyback_value",
    "shares_issued_options", "shares_issued_rsu_vested", "unvested_rsu_shares",
    "antidilutive_shares", "dividends_per_share", "issuance_proceeds",
    # Authorized Headroom
    "shares_authorized", "preferred_shares_authorized",
    "stock_plan_shares_authorized", "buyback_program_authorized",
    # Warrants
    "warrants_outstanding", "warrant_exercise_price", "warrant_shares_callable",
    "warrants_fair_value", "warrant_proceeds",
    # Convertible Debt
    "convertible_debt", "convertible_debt_current",
    "convertible_conversion_price", "convertible_conversion_ratio",
    "convertible_debt_proceeds", "convertible_debt_repayments", "shares_from_conversion",
    # Options Pool
    "options_outstanding", "options_exercisable",
    "options_weighted_avg_price", "options_intrinsic_value",
    # Meta
    "filing_url", "source",
]


def _transform_periods(ticker: str, cik: str, periods: list[dict]) -> pl.DataFrame:
    """Transform extracted periods into a DataFrame matching the financials schema."""
    if not periods:
        return pl.DataFrame()

    rows: list[dict] = []
    for p in periods:
        row: dict = {
            "ticker": ticker,
            "cik": cik,
            "period_start": p.get("period_start"),
            "period_end": p.get("period_end"),
            "fiscal_year": p.get("period_end", "")[:4],
            "fiscal_period": p.get("fiscal_period", ""),
            "form_type": p.get("form_type", ""),
            "filed_date": p.get("filed_date"),
            "accession_number": p.get("accession_number"),
        }

        # Copy all financial fields
        for col in FINANCIALS_COLUMNS:
            if col not in row and col not in ("ticker", "cik", "filing_url", "source"):
                row[col] = p.get(col)

        # Build filing URL
        accn = p.get("accession_number")
        row["filing_url"] = build_filing_url(cik, accn) if accn else ""
        row["source"] = "sec_edgar"

        rows.append(row)

    df = pl.DataFrame(rows, infer_schema_length=None)

    # Cast date columns
    for col in ["period_start", "period_end", "filed_date"]:
        if col in df.columns:
            df = df.with_columns(pl.col(col).cast(pl.Date, strict=False))

    # Ensure all numeric columns are Float64 (avoid int/float mix across rows)
    for col in df.columns:
        if col in ("ticker", "cik", "fiscal_year", "fiscal_period", "form_type",
                   "accession_number", "filing_url", "source",
                   "period_start", "period_end", "filed_date"):
            continue
        if df[col].dtype in (pl.Int64, pl.Int32, pl.UInt64, pl.UInt32):
            df = df.with_columns(pl.col(col).cast(pl.Float64, strict=False))

    return df


def run_financials_backfill(
    tickers: list[str],
    *,
    cik_map: CIKMap | None = None,
    client: httpx.Client | None = None,
) -> int:
    """Backfill financials from SEC EDGAR companyfacts API.

    Args:
        tickers: List of ticker symbols.
        cik_map: Pre-loaded CIK map (created if None).
        client: Pre-created httpx client (created if None).

    Returns:
        Total rows inserted.
    """
    ensure_schema()
    ch = get_client()

    if cik_map is None:
        cik_map = CIKMap()
        cik_map.load()

    own_client = client is None
    if own_client:
        client = make_client()

    start_time = time.monotonic()
    total_rows = 0
    total_periods = 0
    skipped = 0
    failures: list[str] = []

    logger.info("EDGAR financials backfill: %d tickers", len(tickers))

    try:
        for idx, ticker in enumerate(tickers, 1):
            cik = cik_map.cik(ticker)
            if not cik:
                logger.debug("[%d/%d] %s: no CIK mapping, skipping", idx, len(tickers), ticker)
                skipped += 1
                continue

            try:
                data = get_companyfacts(client, cik)
                if not data:
                    logger.debug("[%d/%d] %s: no companyfacts data", idx, len(tickers), ticker)
                    skipped += 1
                    continue

                periods = extract_financials(data)
                if not periods:
                    logger.debug("[%d/%d] %s: no financial periods extracted", idx, len(tickers), ticker)
                    skipped += 1
                    continue

                df = _transform_periods(ticker, cik, periods)
                if df.is_empty():
                    continue

                ch.insert_arrow("financials", df.to_arrow())
                total_rows += len(df)
                total_periods += len(periods)

                if idx % 100 == 0 or idx == len(tickers):
                    logger.info(
                        "[%d/%d] %s: %d periods — total: %s rows",
                        idx, len(tickers), ticker, len(periods), f"{total_rows:,}",
                    )

            except Exception as exc:
                logger.error("[%d/%d] %s: FAILED — %s", idx, len(tickers), ticker, exc)
                failures.append(ticker)

    finally:
        if own_client:
            client.close()

    elapsed = time.monotonic() - start_time
    logger.info(
        "EDGAR financials complete: %s rows, %d periods, %d skipped in %.1f min",
        f"{total_rows:,}", total_periods, skipped, elapsed / 60,
    )
    if failures:
        logger.warning("Failed tickers (%d): %s", len(failures), ", ".join(failures[:50]))

    return total_rows
